[RFC] Node.js clustering in Kibana (#94057)
elastic/kibana: rfcs/text/0015_nodejs_clustering.md

- Start Date: 2021-03-09
- RFC PR: https://github.com/elastic/kibana/pull/94057
- Kibana Issue: https://github.com/elastic/kibana/issues/68626
- POC issue: https://github.com/elastic/kibana/pull/93380

# Summary

Leveraging NodeJS clustering API to have multi-processes Kibana instances.

# Motivation

Kibana currently uses a single Node process/thread to serve HTTP traffic.
This means Kibana cannot take advantage of multi-core hardware making it expensive to scale out Kibana.

> Clustering mode would allow spawning multiple Kibana processes ('workers') from a single Kibana instance.

Reviewer: I think it's maybe worth expanding this section a bit to explain the difference between clustering and nodejs worker threads. Both allows for scaling work to run on multiple core's, but clustering allows for sharing the same network socket so that incoming requests can be shared among workers, whereas the worker threads are more suited for CPU-intensive background work like running the optimizer or maybe encrypting a large file.

In 'classic' mode, the Kibana server is started in the main NodeJS process.

![image](../images/15_clustering/no-cluster-mode.png)

In clustering mode, the main NodeJS process would only start the coordinator, which would then fork workers using 
the cluster API. NodeJS underlying socket implementation allows multiple processes to listen to the same ports, 
performing http traffic balancing between the workers for us.

![image](../images/15_clustering/cluster-mode.png)

The coordinator sole responsibility is to orchestrate the workers. It would not be a 'super' worker handling both 
the job of a worker while being in charge of managing the other workers.

# Performance testing

## Local testing

These tests were performed against a local development machine, with a 8-core cpu (2.4 GHz 8-Core Intel Core i9 - 32 GB 2400 MHz DDR4),
using the default configuration of the `kibana-load-testing` tool.

### Raw results

#### Non-clustered mode

![image](../images/15_clustering/perf-no-clustering.png)

#### Clustered mode, 2 workers

![image](../images/15_clustering/perf-clustering-2-worker.png)

#### Clustered mode, 4 workers

![image](../images/15_clustering/perf-4-workers.png)

> ### Analysis 

Reviewer: Maybe you could explain a bit more how our benchmark test works? My understanding is that they perform a single flow and measure the response times for completing that flow. This is a good smoke test to see what performance a single user would observe on a mostly idle system. However, does our benchmarking run enough parallel requests to start maxing out a single and even more so, multiple core's?

Unless our benchmark is causing the node process to consume so much CPU that we start seeing delays in the event-loop (and some requests start hitting a timeout) we're unlikely to see the true performance gain clustering could achieve.

Although faster response time is a good sign that Kibana is able to handle more, we probably want to measure the request per second throughput.

Author: > Although faster response time is a good sign that Kibana is able to handle more, we probably want to measure the request per second throughput.

Yea, this is not how gatling works tbh. The best we can do would be increasing the number of performed requests per batch.

Reviewer: I don't know anything about our setup but it seems like gattling should be able to give us requests succeeded vs requests failed (OK vs KO) https://gatling.io/docs/current/general/reports/

If we can get this metric we should push concurrent users up until we start seeing at least some failures (like maybe 1-5%).

We would then have to double the concurrent users when trying with two workers and compare the number of successful requests against the one worker scenario.

Reviewer 2: FWIW, when doing performance testing for Fleet, I had to do what @rudolf suggested and manually increase the concurrent users to find the breaking point.

- Add the process info to the log pattern and output it with the log message

We could add the process name information to the log messages, and add a new conversion to be able to display it with the pattern layout, such as %process for example.

the default pattern could evolve to (ideally, only when clustering is enabled)

```
[%date][%level][%process][%logger] %message
```

The logging output would then look like

```
[2021-03-02T10:23:41.834+01:00][INFO ][worker-1][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.840+01:00][INFO ][worker-2][plugins-service] Plugin initialization disabled.
```

This will require some changes in the logging system implementation, as currently the BaseLogger has no logic to 
enhance the log record with context information before forwarding it to its appenders, but this still
seems to be the best solution.

### The rolling-file appender

The rolling process of the `rolling-file` appender is going to be problematic in clustered mode, as it will cause 
concurrency issues during the rolling. We need to find a way to have this rolling process clustered-proof.

Identified options are:

> - have the rolling file appenders coordinate themselves when rolling

Reviewer: nit: can use you numbering for these options so that it's easier to see what you're referring to below

Overall, if no option is trivial, I feel like option 1) is still the most pragmatic one. Option 3) is probably a better 
design, but represents more work. It would probably be fine if all our appenders were impacted, but as only the 
rolling-file one is, I feel like going with 1) would be ok.

### The status API

In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
meaning that it would not be possible to know the status of the cluster as a whole.

We will need to add a centralized status service in the coordinator. Also, as the `/status` endpoint cannot be served 
from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
the status endpoint.

We may also want to have the `/status` endpoint display each individual worker status in addition to the 
global status, which may be breaking change in the `/status` API response format.

### PID file

> Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 

Reviewer: What's the purpose of or how does Kibana use the PID file?

Reviewer 3: @jbudz should know better than me, but my understanding is that it's used by our RPM packages for system service shutdowns?

Reviewer 4: :+1: - in 7.x we support SysV, which reads from the PID file for statuses, [ref](https://github.com/elastic/kibana/blob/7f2afac9972f44cbdb7f335b7b94b6af39e891f0/src/dev/build/tasks/os_packages/service_templates/sysv/etc/init.d/kibana#L106).  The file is generated by the startup script (not using the kibana.yml setting).

We've removed SysV in support in 8.0, so this isn't strictly necessary but still useful to have for external process monitors to attach to.  One example is monit, it will monitor cpu, memory usage, and other metrics and send out alerts if a process is misbehaving.

usage, as the PID stored in the file will be a arbitrary worker’s PID, instead of the coordinator (main process) PID.

In clustering mode, we will need to have to coordinator handle the PID file logic, and to disable pid file handling 
in the worker's environment service.

### SavedObjects migration

In the current state, all workers are going to try to perform the migration. Ideally, we would have only one process 
perform the migration, and the other ones just wait for a ready signal. We can’t easily have the coordinator do it, 
so we would probably have to leverage the ‘main worker’ concept to do so.

The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
Kibana, so this can probably be considered an improvement.

## Open questions and things to solve

### Memory consumption
In cluster mode, node options such as `max-old-space-size` will be used by all processes. E.g using 
`--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 

> If this something we will need to document somewhere?

Author: > Clustering should probably be disabled by default

Yea, it will likely be the case

> In the clustering docs we could explain all the implications of scaling out Kibana.

Yea. even without using any memory option, we still need to document that we clustering mode will spawn multiple processes, with all the underlying implications.

### Workers error handling

When using `cluster`, the common best practice is to have the coordinator recreate workers when they terminate unexpectedly. 
However, given Kibana's architecture, some failures are not recoverable (workers failing because of config validation, failed migration...). 

For instance, if a worker (well, all workers) terminates because of an invalid configuration property, it doesn't make
any sense to have the coordinator recreate them indefinitely, as the error requires manual intervention. 

Should we try to distinguish recoverable and non-recoverable errors, or are we good terminating the main Kibana process 

> when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?

Reviewer 5: In my experience, there are different type of events, and they need to be handled from both ends: 
- Sometimes, the IPC channel is closed due to the host being overloaded. They usually auto-heal, but if the coordinator exits during that disconnection, the worker won't be stopped and we'll get a zombie process. If I'm recalling correctly, that's a `disconnect` event that needs to be handled by the worker to self-kill itself.
- On `exit`, there's the exit code and the [`exitAfterDisconnect`](https://nodejs.org/docs/latest-v14.x/api/cluster.html#cluster_worker_exitedafterdisconnect) flag that could help with identifying if it's a broken process or anything intentional.

But I think it's worth considering what @rudolf says: probably in conjunction with the graceful shutdowns: i.e.: one worker fails, we send the `kill` signal so all the other workers gracefully stop before being killed.

> when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?

Reviewer 6: Current proposal is to kill all processes as Rudolf describes, with worker restarts happening in a follow-up phase after we make a plan for identifying recoverable vs non-recoverable errors.

Kibana currently uses a single Node process/thread to serve HTTP traffic.
This means Kibana cannot take advantage of multi-core hardware making it expensive to scale out Kibana.

Clustering mode would allow spawning multiple Kibana processes ('workers') from a single Kibana instance.

This should be an optional way to run Kibana since users running Kibana inside docker containers might
choose to rather use their container orchestration to run a container (with a single kibana process)per host CPU.

# Architecture summary

In 'classic' mode, the Kibana server is started in the main NodeJS process.

![image](../images/15_clustering/no-cluster-mode.png)

In clustering mode, the main NodeJS process would only start the coordinator, which would then fork workers using 
Node's `cluster` API. NodeJS underlying socket implementation allows multiple processes to listen to the same ports, 
performing http traffic balancing between the workers for us.

![image](../images/15_clustering/cluster-mode.png)

The coordinator sole responsibility is to orchestrate the workers. It would not be a 'super' worker handling both 
the job of a worker while being in charge of managing the other workers.

> # Performance testing

Author: > How much could we save if we did the worker spawning / forking after all of Kibana's source code is loaded/compiled?

The forking just creates a new process with the same entrypoint. We can't 'gain time' by forking later I believe?
Also in production, we are pretranspiling typescript to javascript, so we don't have this `babel/register` overheat. 

>  In theory that should allow for more memory sharing via fork's copy-on-write semantics 

I don't have any knowledge of node's actual implementation to have any idea on that one tbh.

> # Performance testing

Reviewer 2: As far as I'm aware, Node.js's `cluster` module uses `child_process::fork` under the hood, which doesn't use a POSIX `fork` operation:

> Unlike the fork(2) POSIX system call, child_process.fork() does not clone the current process.

Source: https://nodejs.org/api/child_process.html#child_process_child_process_fork_modulepath_args_options

> # Performance testing

Reviewer 3: Got it, that makes more sense then, I was assuming it used POSIX forking.

the improvements of the clustering mode on performance under heavy load are far from negligible.

## Testing against cloud

There is currently no easy way to test the performance improvements this could provide on Cloud, as we can't
deploy custom builds or branches on Cloud at the moment.

On Cloud, Kibana is running in a containerised environment using CPU CFS quota and CPU shares. 

If we want to investigate the potential perf improvement on Cloud further, our only option would be to setup a 
similar-ish environment locally (which wasn't done during the initial investigation)

# Implementation proposal

## enabling clustering mode

Enabling clustering mode will be done using the `clustering.enabled` configuration property.

The whole `clustering` configuration would look like:

```ts
export const config = {
  path: 'clustering',
  schema: schema.object({
    enabled: schema.boolean({ defaultValue: false }),
    workers: schema.number({ defaultValue: 2 }),
  }),
};
```

Notes:
- What should be the default value for `clustering.workers`? We could go with `Max(1, os.cpus().length - 1)`, but do we really want to use all cpus by default,

>   knowing that every worker is going to have its own memory usage.

Reviewer 3: ++ to this question. Also curious what Elasticsearch does here, it's possible we can follow a similar pattern. If we can minimize memory overhead (see comment above), it'd be great to have an sensible automatic scaling default.

>   knowing that every worker is going to have its own memory usage.

Reviewer: The process' RSS memory will grow with the number of open requests. So, theoretically, for a given amount of requests per second, more workers will require less memory each. However, the garbage collector might not collect as aggressively if the RSS is below the maximum heap. And even if there's a trend in our benchmarks it doesn't mean that there won't be spikes, if a given worker handles a request to export a large enough amount of saved objects that worker will consume all of it's heap and adding more workers won't improve this.

So in practice I think we should ignore any memory benchmarks and create the expectation with users that each worker will use up `--max-old-space-size` (or the default for their system).

>   knowing that every worker is going to have its own memory usage.

Reviewer: > do we really want to use all cpus by default

It feels like there really isn't a good way to automatically choose the correct value here. Perhaps we should just always require that the user sets this in configuration if they want clustering enabled?

```ts
type ClusterMessagePayload = Serializable;

interface BroadcastOptions {
  /**
   * If true, will also send the message to the worker that sent it.
   * Defaults to false.
   */
  sendToSelf?: boolean;
  /**
   * If true, the message will also be sent to subscribers subscribing after the message was effectively sent.
   * Defaults to false.
   */
  persist?: boolean;
}

export interface ClusteringServiceSetup {
  /**
   * Return true if clustering mode is enabled, false otherwise
   */
  isEnabled: () => boolean;
  /**
   * Return the current worker's id. In non-clustered mode, will return `1`
   */
  getWorkerId: () => number;
  /**
   * Broadcast a message to other workers.
   * In non-clustered mode, this is a no-op
   */
  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
  /**
   * Registered a handler for given `type` of IPC messages
   * In non-clustered mode, this is a no-op that returns a no-op unsubscription callback.
   */
  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;

>   /**
>    * Returns true if the current worker has been elected as the main one. In non-clustered mode, will return true
>    */

Reviewer 3: nit
```suggestion
  /**
   * Returns true if the current worker has been elected as the main one. In non-clustered mode, will always return true
   */
```

  isMainWorker: () => boolean;
}
```

## Cross-worker communication

For some of our changes (such as the /status API, see below), we will need some kind of cross-worker communication. This 
will need to pass through the coordinator, which will also serve as an 'event bus', or IPC forwarder.

This IPC API will be exposed from the clustering service.

```ts
export interface ClusteringServiceSetup {
  // [...]
  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;

>   addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;

Author: I was wondering the same thing, I even added the question as a note a few lines down :)

}
```

Notes:
-  to reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 
   It will avoid to force (most) code to check in which mode Kibana is running before calling them.
-  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.

## Executing code on a single worker

In some scenarios, we would like to have parts of the code executed only from a single process. 

The SO migration would be a good example: 
we don't need to have each worker try to perform the migration, and we'd prefer to have one performing/trying the migration, 
and the others wait for it. Due to the architecture, we can't have the coordinator perform such single-process jobs, 
as it doesn't actually run a Kibana server.

> There are various ways to address such use-cases. What seems to be the best compromise right now would be the concept of 

Author: Yea, the main worker will be a worker like the others, the only distinction will be this flag, that would allow plugins to execute arbitrary parts of the code only on a single worker if they need to.

'main worker'. The coordinator would arbitrary elect a worker as the 'main' one at startup. The clustering service would then expose 

> an API to let workers identify themselves as main or not.

Reviewer 3: IMO, a main worker per plugin seems like a high cost of complexity. I think it's much simpler to reason about there being a single main process rather than distributing this responsibility across the workers on a per-plugin basis.

I hope that we can find a way for things like Task Manager and Telemetry to operate correctly on multiple workers within the cluster, just as it does today on multiple instances.

> an API to let workers identify themselves as main or not.

Author: > IMO, a main worker per plugin seems like a high cost of complexity

Yea, also, we can't be sure that code executed by the 'main' worker from a given plugin doesn't need something done by the same main worker by another plugin (even if that seems terrible design).

> Otherwise, it feels like clustering would only help with HTTP requests because all the other background processes will likely require one 'main' node for the sake of avoiding concurrency issues.

In the initial implementation, maybe. Even if I don't think background tasks are really a significant part of our load, especially under heavy load (which is what clustering want to address).

More longer term, plugins should adapt their implementation to add inter-worker communication. But we can't realistically have all this done in the initial implem, which is what this main worker concept tries to address.

```ts
export interface ClusteringServiceSetup {
  // [...]
  isMainWorker: () => boolean;
}
```

Notes:
- in non-clustered mode, `isMainWorker` would always return true, to reduce the divergence between clustered and 
  non-clustered modes.

### Example: SO Migration

To take the example of SO migration, the `KibanaMigrator.runMigrations` implementation could change to
(naive implementation, the function is supposed to return a promise here, did not include that for simplicity):

```ts
runMigration() {
   if(clustering.isMainWorker()) {
     this.runMigrationsInternal().then((result) => {
        applyMigrationState(result);
        clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent
      })
   }
   else {
     const unsubscribe = clustering.addMessageHandler('migration-complete', ({ payload: result }) => {
       applyMigrationState(result);
       unsubscribe();

>      });

Reviewer 3: We may want some guard rails around this message handling API. For example, I think it would be best if we segmented these messages so that plugins can only read messages that come from the same plugin to avoid any implicit cross-plugin APIs.

   }
}
```

Notes:
  - to be sure that we do not encounter a race condition with the event subscribing / sending (workers subscribing after 
    the main worker actually send the `migration-complete` event and then waiting indefinitely), we are using the `persist` option of the `broadcast` API.
    I'm honestly not sure this is the best pattern, but I couldn't come with anything better (maybe shared state instead?). 

>     This is probably something else we should discuss.

Reviewer 3: I prefer the `persist` option on messages over using shared state. Main reason being that we can easily log it and trace the order of events & execution to debug issues that arise.

If we do need shared state, we will probably have to use syscall libraries to share buffers such as [mmap-io](https://www.npmjs.com/package/mmap-io), 
and expose a higher level API for that from the `clustering` service.

# Technical impact on the codebase

This section is meant to be an exhaustive (even if realistically not at the moment) inventory  of the changes that would
be required to support clustering mode.

## Technical impact on Core

### Handling multi-process logs

This is an example of log output in a 2 workers cluster, coming from the POC:

```
[2021-03-02T10:23:41.834+01:00][INFO ][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.840+01:00][INFO ][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.900+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
```

> The workers logs are interleaved, and, most importantly, there is no way to see which process each log entry is coming from. 

Reviewer 7: Sharing my past experience

As you might be already aware, Without a centralised co-ordinator, the logging to the stdout or any stream would be interleaved this making the output in a broken state if the application becomes chatty or busy. 

One thing that worked very well was to differentiate the master vs the worker mode and use different stream which would eventually write everything to `process.stdout` by subscribing to worker messages. Here is a dummy implementation that worked well

```js
class WorkerStream extends stream.Writable {
    constructor(callback) {
        super({ decodeStrings: false });
        this.callback = callback;
    }

    _write (chunk, _, done) {
        this.callback(chunk);
        done();
    }
}

const workerStream = new WorkerStream((data) => {
    // send to cordinator for logging
    process.send('log', data)
});

function createLogger () {
    const stream = cluster.isWorker ? workerStream : process.stdout
    // abstracted out - could be BaseLogger for Kibana
    return logger(stream)
}

// somewhere in coordinator
process.on('log', data => {
    process.stdout.write(data)
})
```

I don't know much about the Kibana architecture, just sharing my experience working with cluster modules.

Author: > As you might be already aware, Without a centralised co-ordinator, the logging to the stdout or any stream would be interleaved this making the output in a broken state if the application becomes chatty or busy

I wonder to which extend this should be considered blocker tbh. Logs from a given process/worker will still be assured to be in order, we'll have the timestamp and the workerId in the entries, and even if flushes between processed are not assured to be in order, the timegap, even under heavy load, is not that significant.

> One thing that worked very well was to differentiate the master vs the worker mode and use different stream which would eventually write everything to process.stdout by subscribing to worker messages

Yea, I listed that in our options. This feels like the proper design. The only problem is that the overall complexity of implementing that is quite high.

> That also addresses the problem with the rolling files. Doesn't it?

Yea, it definitely will. That's option `3.` of the rolling file appender section.

Reviewer 5: I see more than a sorting/ordering problem, an issue when 2+ processes are writing to the same file. They can mixin entries and have broken logs with some content breaking in the middle of some others. 
I think we definitely need the coordinator to ensure log entries are written one at a time.

Reviewer 6: Current proposal is to centralize logging in the coordinator, which should resolve issues of interleaving as the coordinator would be the only process writing to a file or stdout.

We will need to address that.

Our options are:

- Have a distinct logging configuration for each worker

We could do that by automatically adding a suffix depending on the process to the appropriate file appenders configuration. 
Note that this doesn’t solve the problem for the console appender, which is probably a no-go.

- Add the process info to the log pattern and output it with the log message

We could add the process name information to the log messages, and add a new conversion to be able to display it with 

> the pattern layout, such as `%process` for example.

Reviewer 3: FYI we already have support for `%pid` in the pattern layout, but it is not included in the default. I think it may be better to name this `%worker` in order to differentiate from process ID.

Author: Yea, I wanted to name it with something that sounds more specific to the clustering API, however we will likely also have the coordinator output logs, so I wasn't really sure about `%worker`. But that be fine anyway

the default pattern could evolve to (ideally, only when clustering is enabled)

```
[%date][%level][%process][%logger] %message
```

The logging output would then look like

```
[2021-03-02T10:23:41.834+01:00][INFO ][worker-1][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.840+01:00][INFO ][worker-2][plugins-service] Plugin initialization disabled.
```

This will require some non-trivial changes in the logging system implementation, as currently the BaseLogger has no logic to 
enhance the log record with context information before forwarding it to its appenders, but this still
seems to be the best solution. (This may even be the first step to implement our MDC)

Notes:

> - Even if we add the `%process` pattern, do we want to also allow users to specify per-worker log files? 

Reviewer 3: Do we have any ideas on why this would a be useful feature? Seems like this can easily be emulated with `tail -f my.log | grep 'worker-1'` or filtering on the ECS logs within Elasticsearch.

Author: I personally don't really see any proper use case, but I still added the question as someone else might.

### The rolling-file appender

The rolling process of the `rolling-file` appender is going to be problematic in clustered mode, as it will cause 
concurrency issues during the rolling. We need to find a way to have this rolling stage clustered-proof.

Identified options are:

- have the rolling file appenders coordinate themselves when rolling

By using a broadcast message based mutex mechanism, the appenders could acquire a ‘lock’ to roll a specific file, and

> notify other workers when the rolling is complete (quite similar to what we want to do with SO migration for example).

Reviewer 3: I'm not sure what our guarantees are on how exact our log rolling is, but if we relax the requirement of rolling happening exactly at a given file size to allow some margin of error, we could just have only the `main` worker do the rolling to avoid any needs for locking and then send a message to other workers to reopen the file handle. This could result in some log messages getting written to the old file after it's renamed, but maybe that'd be ok?

Author: That could be a slightly simpler alternative to option 1. yea, as it would reduce the inter-worker communication to notification when the main worker has finished the rolling. I think loosing control on the exact size the file will have when rolling should be acceptable.

- have the coordinator process perform the rolling

Another option would be to have the coordinator perform the rotation instead. When a rolling is required, the appender 
would send a message to the coordinator, which would perform the rolling and notify the workers once the operation is complete. 

Note that this option is even more complicated than the previous one, as it forces to move the rolling implementation 
outside of the appender, without any significant upsides identified.

> - centralize the logging system in the coordinator (and have the workers send messages to the coordinator when using the logging system)

Reviewer 3: Would we have to centralize it in the coordinator process? Wouldn't it be simpler to just forward the logs to the `main` Kibana worker? That way we wouldn't have to two very different implementations or pull any logging code outside of Core, but can still benefit from having a single process doing the writing.

Author: > That way we wouldn't have to two very different implementations or pull any logging code outside of Core

I don't think we need to pull logging code outside of core in either option? The coordinator class belongs to Core too.

Regarding the implementations, both approaches have the same complexity I believe, as both requires to short circuit the current / default logging system implementation to send messages instead of invoking the logging logic when in clustering mode.

Also, the coordinator will probably need to also output logs, and it would be awkward to have the coordinator rely on the main worker to process its own logs I think?

We could go further, and change the way the logging system works in clustering mode by having the coordinator centralize 
the logging system. The worker’s logger implementation would just send messages to the coordinator. If this may be a 
correct design, the main downside is that the logging implementation would be totally different in cluster and 
non cluster mode, and seems to be way more work that the other options.

Overall, if no option is trivial, I feel like option 1) is still the most pragmatic one. Option 3) may be a better 
design, but represents more work. It would probably be fine if all our appenders were impacted, but as only the 
rolling-file one is, I feel like going with 1) would be ok.

### The status API

In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 

> meaning that it would not be possible to know the status of the cluster as a whole.

Author: I would assume (well, guess) that the amount of 'background' / 'system' requests such license fetching or ES healthchecks should be negligible compared to the amount of user performed requests?

Kibana, so this can probably be considered an improvement.

## Open questions and things to solve

### Memory consumption

In clustered mode, node options such as `max-old-space-size` will be used by all processes. 

E.g using `--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 

If this something we will need to document somewhere?

### Workers error handling

When using `cluster`, the common best practice is to have the coordinator recreate ('restart') workers when they terminate unexpectedly. 
However, given Kibana's architecture, some failures are not recoverable (workers failing because of config validation, failed migration...). 

For instance, if a worker (well, all workers) terminates because of an invalid configuration property, it doesn't make
any sense to have the coordinator recreate them indefinitely, as the error requires manual intervention. 

Should we try to distinguish recoverable and non-recoverable errors, or are we good terminating the main Kibana process 
when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?

### Data folder

> The data folder (`path.data`) is currently the same for all workers. 

Reviewer 4: +1.  There's the packages/apm-config-loader instance for fetching UUID.  keystore for writes.  So far no plugin instances that I know of.

We still have to identify with the teams if this is going to be a problem. It could be, for example, if some plugins
are accessing files in write mode, which could result in concurrency issues between the workers.

If that was confirmed, we would have to create and use a distinct data folder for each worker.

One pragmatic solution could be, when clustering is enabled, to create a sub folder under path.data for each worker. 

If we need to do so, should this be considered a breaking change, or do we assume the usage of the data folder is an 

> implementation detail and not part of our public 'API'?

Reviewer 3: I don't believe we consider this part of our public API and I'm pretty sure we've changed this in the past during minor releases.

### instanceUUID

The same instance UUID (`server.uuid` / `{dataFolder}/uuid`) is currently used by all the workers. 

We still need to identify with the teams if this is going to be a problem, in which case we would need to have distinct 
instance uuid per workers.

Note that this could be a breaking change, as the single `server.uuid` configuration property would not be enough. 

> Could the new `clustering.getWorkerId()` API be used here somehow? We could have a unique worker id with `serverUUid-workerId`. 

Reviewer 3: Yeah I was going to suggest this same `serverUUid-workerId` pattern. One key benefit is that provides a more visible migration path for turning on clustering since the base UUID will not change for existing deployments (assuming a persistent disk).

### The Dev CLI

In development mode, we are already spawning processes from the main process: The Kibana server running in the main process 
actually just kickstarts core to get the configuration and services required to instantiate the `CliDevMode` 

> (from the legacy service), which itself spawns a child process to run the actual Kibana server.

Reviewer 3: @mistic can Bazel be of any aid with this? I think that once Bazel is handling the webpack builds, the only thing we really need multiple processes for in the dev CLI is for base path proxying. It'd be neat if we could create a Bazel target that starts up an nginx instance for base path proxying and a Kibana server process (w/ clustering enabled if desired).

Author: This will be addressed by https://github.com/elastic/kibana/issues/76935, which will very likely land before any implementation attempt for this proposal. I wonder if I should just remove this section?

Reviewer 6: removed

- Using instanceUUID as a unique Kibana process identifier

Is there, for example, schedulers that are using the instanceUUID a single process id, in opposition to a single 
Kibana instance id? Are there situations where having the same instance UUID for all the workers is going to be a problem?

- Things needing to run only once per Kibana instance

Is there any part of the code that needs to be executed only once in a multi-worker mode, such as initialization code, 
or starting schedulers? 

An example would be reporting's queueFactory pooling. As we want to only be running a single headless at a time per 
Kibana instance, only one worker should have pooling enabled.

### Identified required changes

(Probably not exhaustive at the moment)

#### Reporting

We will probably want to restrict to a single headless per Kibana instance. For that, we will have to change the logic 
in [createQueueFactory](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/reporting/server/lib/create_queue.ts#L60-L64) 
to only have the 'main' worker be pooling for reporting tasks.

> #### Telemetry

Reviewer 5: Event-based telemetry may be affected as well. Both: the existing one in the Security Solutions team (cc @stevewritescode), and the general one that is in the works.

More specifically, the size of the queues will be multiplied per worker, also growing in the amount of network bandwidth used, and potentially affecting our customers.

We could make sure that the queues are held only in the `main` node.

- Server side fetcher
  
The telemetry/server/fetcher.ts will attempt sending the telemetry usage multiple times once per day from each process. 
We do store a state in the SavedObjects store of the last time the usage was sent to prevent sending multiple times 
(although race conditions might occur).

- Tasks storing telemetry data
  
We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
tasks will be registered multiple times.

> Note that sending the data multiple times doesn’t have any real consequences, so this should be considered 

Reviewer 5: ```suggestion
Note that sending the data multiple times doesn’t have any real consequences apart from the additional number of ES requests, so this should be considered 
```

status API in clustering mode. 

Note that the new format for /api/status is still behind a v8format flag, meaning that if we do these changes before 8.0, 
we won't be introducing any breaking change later.

## instanceUUID

Depending on our decision regarding the instanceUUID problematic, we may have to change the server.uuid configuration 
property when clustering mode is enabled, which would be a breaking change.

## distinct logging configuration

If we decide to have each worker output log in distinct files, we would have to change the logging configuration to 
add a prefix/suffix to each log file, which would be a breaking change. Note that this option is probably not the one 
we'll choose.

## data folder

We may need to have each worker uses a distinct data folder. Should this be considered a breaking change?

# Drawbacks

- Implementation cost is going to be significant, both in core and in plugins. Also, this will have to be a collaborative
  effort, as we can't enable the clustered mode in production until all the identified breaking changes have been addressed.
  

> - Even if easier to deploy, it doesn't really provide anything more than a multi-instances Kibana setup.

Reviewer 3: Do we have any side-by-side comparisons of the performance of two modes?

Reviewer: > it doesn't really provide anything more than a multi-instances Kibana setup.

Kibana does a lot of background polling work which creates a significant amount of network traffic. If we centralize all this background polling to a single worker we can reduce this background work by a factor equal to the number of worker processes (maybe that would be a factor of 8 or 16 for most deployments?).

It's probably worth trying to quantify the impact, e.g. reducing 1Kb of traffic per month by a factor of 16 is negligible but reducing 10Gb by a factor of 16 would be a big win.

Author: > If we centralize all this background polling to a single worker we can reduce this background work by a factor equal to the number of worker processes

That sounds great, but would also require more changes from plugins though.

E.g we could have only one worker perform the license check calls, and then broadcast the result to the other workers, but that means changing the `licensing` plugin implementation.

All these optimization can easily be done later as follow-ups though.

> It's probably worth trying to quantify the impact

We would need to list all the 'background' requests we are performing against ES. Not sure how to do that without help from the other teams though.

Reviewer: Agreed, but I think it's worth adding this as a benefit to clustering.

Reviewer 6: This is listed as a potential future benefit of clustering, and the current RFC doesn't preclude us from doing this. However the current proposal is to treat it as an enhancement and not address in the initial implementation.

  
- This will complexify the code, especially in Core were some part of the logic will drastically change between clustered and 
non-clustered modes (e.g the status API, rolling-file appender).
  
- It could introduce subtle bugs in clustered mode, as we may overlook some breaking changes, or developers may omit to
ensure clustered mode compatibility when adding new features.
  

> - Proper testing of all the edge cases is going to be tedious, if not just realistically impossible. We can't really

Author: Well, the most significant problem imho in clustering testing is to ensure that no part of the code is doing that are considered required changes (e.g concurrent write to a file). But I don't really see how you can effectively ensure that no part of the code is doing that using automated testing, especially when the issues are often very hard to reproduce (e.g races)

we'll choose.

# Drawbacks

- Implementation cost is going to be significant, both in core and in plugins. Also, this will have to be a collaborative
  effort, as we can't enable the clustered mode in production until all the identified breaking changes have been addressed.
  
- Even if easier to deploy, it doesn't really provide anything more than a multi-instances Kibana setup.
  
- This will complexify the code, especially in Core were some part of the logic will drastically change between clustered and 
non-clustered modes (e.g the status API, rolling-file appender).
  
- It could introduce subtle bugs in clustered mode, as we may overlook some breaking changes, or developers may omit to
ensure clustered mode compatibility when adding new features.
  
- Proper testing of all the edge cases is going to be tedious, if not just realistically impossible. We can't really
  automate the testing of the clustered mode.

# Alternatives

One alternative of the `cluster` module is using a worker pool via `worker_threads`. Both have distinct use cases
though. Clustering is meant to have multiple workers with the same codebase, often sharing a network socket to balance

> network traffic. The worker pattern is a way to create specialized workers in charge of executing isolated, CPU intensive

Reviewer: nit: it's not clear which one of the two "worker pattern" refers to

```suggestion
network traffic. Worker threads is a way to create specialized workers in charge of executing isolated, CPU intensive
```

Even if increasing the number of workers doesn’t just linearly increase the performances 
(which totally make sense, most of our requests response time is caused by awaiting ES response), 
the improvements of the clustering mode on performance under heavy load are far from negligible.

## Testing against cloud

There is currently no easy way to test the performance improvements this could provide on Cloud, as we can't
deploy custom builds or branches on Cloud at the moment.

On Cloud, Kibana is running in a containerised environment using CPU CFS quota and CPU shares. 

If we want to investigate the potential perf improvement on Cloud further, our only option would be to setup a 
similar-ish environment locally (which wasn't done during the initial investigation)

# Implementation proposal

## enabling clustering mode

Enabling clustering mode will be done using the `clustering.enabled` configuration property.

The whole `clustering` configuration would look like:

```ts
export const config = {
  path: 'clustering',
  schema: schema.object({
    enabled: schema.boolean({ defaultValue: false }),
    workers: schema.conditional(
      schema.siblingRef('enabled'),
      false,

>       schema.number({ defaultValue: 1 }),

Reviewer 2: If this is set to 1, we will have a coordinator and a single worker, right? That wouldn't get us that much more CPU usage than not having clustering mode at all. Should this default be 2?

Reviewer 6: This default would just be applied if clustering.enabled is set to false as it is wrapped in `conditional()`, but i've removed this section anyway and replaced it with yaml for simplicity

      schema.number()
    )
  }),
};
```

Notes:

> - As there isn't really a good way to automatically choose the correct number of workers automatically, It will be required 

Reviewer 2: Is it possible to make this default to the number of CPUs available?

Author: It would be, but from previous discussions we agreed that this could be a potentially dangerous behavior, which is why we decided to force the user to manually specify the value:
- There is the coordinator process to take into account, so the 'ideal' value would be numCpu-1
- We're not sure what else is running on the machine. Having a process per CPU per default could be a potentially dangerous default

Reviewer 6: Updated to reflect a strategy of using all available CPUs (- 1 for the coordinator)

   */
  isEnabled: () => boolean;
  /**
   * Return the current worker's id. In non-clustered mode, will return `1`
   */
  getWorkerId: () => number;
  /**
   * Broadcast a message to other workers.
   * In non-clustered mode, this is a no-op
   */
  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
  /**
   * Registered a handler for given `type` of IPC messages
   * In non-clustered mode, this is a no-op that returns a no-op unsubscription callback.
   */
  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
  /**
   * Returns true if the current worker has been elected as the main one. In non-clustered mode, will always return true
   */
  isMainWorker: () => boolean;
}
```

### Example: SO Migration

To take the example of SO migration, the `KibanaMigrator.runMigrations` implementation could change to
(naive implementation, the function is supposed to return a promise here, did not include that for simplicity):

```ts
runMigration() {
   if(clustering.isMainWorker()) {
     this.runMigrationsInternal().then((result) => {
        applyMigrationState(result);

>         clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent

Author: > If we used Observables as the emit and subscribe interface, I don't believe we would need this persist option

It mostly depends on the underlying observable implementation we choose to implement. We need to keep in mind that if we use a observable pattern, we will need to code the implementation ourselves, as the actual IPC protocol we're using is event-based. So we'll be just creating an observable from the IPC messages exposed by the `clusteringService` on the workers side. 

If we get rid of this `persist` option and choose to use observable, and decide than a new subscriber should receive all previous events, we will then need to keep ALL previous IPC messages in memory, which could have some serious memory impact on a long-running Kibana instance.

>  For instance, if there is some sort of shared state that needs to be broadcast, it'd be nice to use a ReplaySubject(1)

This would not solve the synchronisation issues, as, unless we do store all messages to re-emit them, the `replay` effect would still have to be applied somewehere in the chain **before** the actual emission of the initial event go through said chain, which, in given example, is not necessarily true depending on processes race.  

So yea, we probably do want an Obs-based API, but that doesn't quite solve the technical challenges of pub/sub using IPC.

>         clustering.broadcast('migration-complete', { payload: result }, { persist: true }); // persist: true will send message even if subscriber subscribe after the message was actually sent

Reviewer 6: I added observable-based API to the RFC as a future enhancement to investigate, as described in my comments below.

be required to support clustering mode.

## Technical impact on Core

### Handling multi-process logs

This is an example of log output in a 2 workers cluster, coming from the POC:

```
[2021-03-02T10:23:41.834+01:00][INFO ][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.840+01:00][INFO ][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.900+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
```

The workers logs are interleaved, and, most importantly, there is no way to see which process each log entry is coming from. 
We will need to address that.

Our (non-exclusive) options are:

1. Have a distinct logging configuration for each worker

We could do that by automatically adding a suffix depending on the process to the appropriate file appenders configuration. 

> Note that this doesn’t solve the problem for the console appender, which is probably a no-go.

Reviewer 3: > Note that this doesn’t solve the problem for the console appender, which is probably a no-go.

Any reason we couldn't handle these two appenders differently without having to add much logging logic in the coordinating process? It seems the console appender could simply forward the rendered layout output strings to the coordinating process without the coordinating process needing to know any logging configuration.

If this would work, I lean towards this option since it would be the least technical effort, however I'm not sure how common this type of pattern is. Elasticsearch does not appear to handle logging in this way and is using a single file per node (however they are using threads rather than processes).

From a support perspective, I imagine it's simpler if we use a single file. Requesting logs when there are multiple outputs may be confusing for customers.

Reviewer 6: We decided to centralize logging in the coordinator, so this is no longer a concern.

> 2. Add the process info to the log pattern and output it with the log message

Reviewer 6: Decided we'll have the coordinator handle writing log entries, mitigating this issue

We could add the process name information to the log messages, and add a new conversion to be able to display it with 
the pattern layout, such as `%worker` for example.

the default pattern could evolve to (ideally, only when clustering is enabled)

```
[%date][%level][%worker][%logger] %message
```

The logging output would then look like

```
[2021-03-02T10:23:41.834+01:00][INFO ][worker-1][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.840+01:00][INFO ][worker-2][plugins-service] Plugin initialization disabled.
```

This will require some non-trivial changes in the logging system implementation, as currently the BaseLogger has no logic to 
enhance the log record with context information before forwarding it to its appenders, but this still
seems to be the best solution. (This may even be the first step to implement our MDC)

Notes:
- The coordinator will probably need to output logs too. `%worker` would be interpolated to `coordinator` 
  for the coordinator process
- Even if we add the `%worker` pattern, do we want to also allow users to specify per-worker log files? 

> ### The rolling-file appender

Reviewer 6: Decided we'll have the coordinator handle writing log entries, mitigating this issue

notify other workers when the rolling is complete (quite similar to what we want to do with SO migration for example).

An alternative to this option would be to only have the main worker handle the rolling logic. We will lose control
on the exact size the file is when rolling, as we would need to wait until the main worker receives a log message
for the rolling appender before the rolling is effectively performed. The upside would be that it reduces the inter-workers
communication to a notification from the main worker to the others once the rolling is done for them to reopen their
file handler.

2. have the coordinator process perform the rolling

Another option would be to have the coordinator perform the rotation instead. When a rolling is required, the appender 
would send a message to the coordinator, which would perform the rolling and notify the workers once the operation is complete. 

Note that this option is even more complicated than the previous one, as it forces to move the rolling implementation 
outside of the appender, without any significant upsides identified.

> 3. centralize the logging system in the coordinator

Reviewer 2: I think this is going to be our only option if we want to write the logs to a single file or stdout without fear of the messages interleaving.

We could go further, and change the way the logging system works in clustering mode by having the coordinator centralize 
the logging system. The worker’s logger implementation would just send messages to the coordinator. If this may be a 
correct design, the main downside is that the logging implementation would be totally different in cluster and 
non cluster mode, and seems to be way more work that the other options.

Overall, if no option is trivial, I feel like option `1.` is still the most pragmatic one. Option `3.` may be a better 
design, but represents more work. It would probably be fine if all our appenders were impacted, but as only the 
rolling-file one is, I feel like going with `3.` would be ok.

### The status API

In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
meaning that it would not be possible to know the status of the cluster as a whole.

> We will need to add a centralized status service in the coordinator. Also, as the `/status` endpoint cannot be served 

Reviewer 2: Will the coordinator serve any HTTP requests, or will they all be served by the workers themselves?

Without changes, each worker is going to try to write and read the same PID file. Also, this breaks the whole pid file 
usage, as the PID stored in the file will be a arbitrary worker’s PID, instead of the coordinator (main process) PID.

In clustering mode, we will need to have to coordinator handle the PID file logic, and to disable pid file handling 
in the worker's environment service.

### SavedObjects migration

In the current state, all workers are going to try to perform the migration. Ideally, we would have only one process 
perform the migration, and the other ones just wait for a ready signal. We can’t easily have the coordinator do it, 
so we would probably have to leverage the ‘main worker’ concept here.

The SO migration v2 is supposed to be resilient to concurrent attempts though, as we already support multi-instances 
Kibana, so this can probably be considered an improvement.

## Open questions and things to solve

### Memory consumption

In clustered mode, node options such as `max-old-space-size` will be used by all processes. 

E.g using `--max-old-space-size=1024` in a 2 workers cluster would have a maximum memory usage of 3gb (1 coordinator + 2 workers). 

> If this something we will need to document somewhere?

Reviewer 2: It'd be nice if we had the flexibility to allocate different maximum memory usage for the coordinator versus the workers. They're conceptually performing different functionality, so it'd be great to have that level of granularity.

@tylersmalley how do we currently recommend that users adjust the `max-old-space-size`?

Reviewer 8: Currently this is defined in `config/node.options`

Reference: https://www.elastic.co/guide/en/kibana/current/production.html#memory

Author: > It'd be nice if we had the flexibility to allocate different maximum memory usage for the coordinator versus the workers

Not sure if we can do that using the `cluster` API. 

From the doc:, we can only pass env variables when forking

```ts
cluster.fork([env])
```

Reviewer 6: Discussed this below:

> the solution we'd like to propose is to disable clustering if a user has `max-old-space-size` set at all, which would ensure it wouldn't be possible to hit unpredictable behavior. To opt-in for clustering, the user would update their config & remove `max-old-space-size` settings. And if no `max-old-space-size` is set, we could still consider enabling clustering by default once we are confident it is stable.

> If we still want to provide users with fine-grained control over memory-per-worker, we could update the config so users could provide a list of workers instead of a fixed number, perhaps something like:

```yaml
clustering:
  enabled: true
  workers:
    - name: foo # could perhaps use name in the logs as well for extra clarity
      max_old_space_size: 1234
    - name: bar
      max_old_space_size: 1234
```

However, given Kibana's architecture, some failures are not recoverable (workers failing because of config validation, failed migration...). 

For instance, if a worker (well, all workers) terminates because of an invalid configuration property, it doesn't make
any sense to have the coordinator recreate them indefinitely, as the error requires manual intervention. 

Should we try to distinguish recoverable and non-recoverable errors, or are we good terminating the main Kibana process 
when any worker terminates unexpectedly for any reason (After all, this is already the behavior in non-cluster mode)?

### Data folder

The data folder (`path.data`) is currently the same for all workers. 

We still have to identify with the teams if this is going to be a problem. It could be, for example, if some plugins
are accessing files in write mode, which could result in concurrency issues between the workers.

If that was confirmed, we would have to create and use a distinct data folder for each worker.

One pragmatic solution could be, when clustering is enabled, to create a sub folder under path.data for each worker. 

> The data folder is not considered part of our public API, and the implementation and path already changed in previous

Reviewer 2: Really? How did it change in previous minor releases?

Author: This statement is coming from @joshdover, I'll let him answer if he remembers

Reviewer 3: I'm not sure what I said and I'm having a hard time finding an instance where we changed this, I must have misspoke!

However, I don't believe we consider the _structure_ of the data folder to be part of our public API. We should be able to use the sub-folder approach if needed I believe.

Reviewer 2: What all is stored in the data folder? Is it just the `uuid` file? The docs are not really clear...

>
path.data: | The path where Kibana stores persistent data not saved in Elasticsearch. Default: data
-- | --

Source: https://www.elastic.co/guide/en/kibana/current/settings.html

And then they're contradicted by the package directory layout
> 

data | The location of the data files written to disk by Kibana and its plugins | /var/lib/kibana
-- | -- | --
path.data | logs | Logs files location | /var/log/kibana
path.logs | plugins | Plugin files location. Each plugin will be contained in a subdirectory. | /usr/share/kibana/plugins

Source: https://www.elastic.co/guide/en/kibana/current/rpm.html#rpm-layout

Reviewer 3: The only plugin that I have found that is referencing this is the reporting plugin, however I can't find a place in code the actually reads the configuration. @elastic/kibana-reporting-services Could you provide any guidance into how the `path.data` config is used by the Reporting plugin? My guess is as some sort of temporary storage, however I can't find any usages.

Reviewer 6: Pinging @elastic/kibana-app-services for any guidance on what reporting does here.

Reviewer 9: I think we used to use it for the "user data directory" for Chromium, but as I look in the code now, we just use `tmpdir` from the `os` package: https://github.com/elastic/kibana/blob/master/x-pack/plugins/reporting/server/browsers/chromium/driver_factory/index.ts#L51

planned for 7.13 or early 7.14, see [the associated GH issue](https://github.com/elastic/kibana/issues/76935)

## Technical impact on our plugins codebase

### Identifying things that may break

- Concurrent access to the same resources

Is there, for example, some part of the code that is accessing and writing files from the data folder (or anywhere else) 
and makes the assumption that it is the sole process actually writing to that file?

- Using instanceUUID as a unique Kibana process identifier

Is there, for example, schedulers that are using the instanceUUID a single process id, in opposition to a single 
Kibana instance id? Are there situations where having the same instance UUID for all the workers is going to be a problem?

- Things needing to run only once per Kibana instance

Is there any part of the code that needs to be executed only once in a multi-worker mode, such as initialization code, 
or starting schedulers? 

An example would be reporting's queueFactory pooling. As we want to only be running a single headless at a time per 
Kibana instance, only one worker should have pooling enabled.

### Identified required changes

(Probably not exhaustive at the moment)

> #### Reporting

Reviewer 2: Starting in 8.0, Reporting will be using TaskManager.

to only have the 'main' worker be pooling for reporting tasks.

#### Telemetry

- Server side fetcher
  
The telemetry/server/fetcher.ts will attempt sending the telemetry usage multiple times once per day from each process. 
We do store a state in the SavedObjects store of the last time the usage was sent to prevent sending multiple times 
(although race conditions might occur).

- Tasks storing telemetry data
  
We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
tasks will be registered multiple times.

Note that sending the data multiple times doesn’t have any real consequences, apart from the additional number of ES requests, 
so this should be considered non-blocking and only an improvement.

- Event-based telemetry

Event-based telemetry may be affected as well. Both the existing one in the Security Solutions team and the general 
one that is in the works. More specifically, the size of the queues will be multiplied per worker, also growing in the
amount of network bandwidth used, and potentially affecting our customers.

We could address that by making sure that the queues are held only in the main worker.

> #### TaskManager

Reviewer 2: We've started to see quite a few issues with TaskManager running in the same process as the HTTP Server. For example, there was recently a SIEM detection rule that exerted such a load on Kibana that all HTTP requests to Kibana failed.

Long-term, it'd be great to get the TaskManager in its own worker so it can't affect anything else negatively. Will the main worker only initially be responsible for task-manager and running saved-object migrations on startup? If so, it's probably fine to have the TaskManager here short-term.

Reviewer 3: > Long-term, it'd be great to get the TaskManager in its own worker so it can't affect anything else negatively.

I'd take this one step further and say it'd also be great if a single task couldn't starve the other tasks for resources as well. I think to solve this, we'd need to use a proper pool of worker threads to execute individual tasks on.

Agreed that having task manager run on a single worker for now would get us quite a bit of value. At least we can still continue accepting requests when a CPU-intensive task is running.

Reviewer 2: > I'd take this one step further and say it'd also be great if a single task couldn't starve the other tasks for resources as well. I think to solve this, we'd need to use a proper pool of worker threads to execute individual tasks on.

👍

Reviewer 6: This has been clarified in the RFC. Current proposal is for TM to live in the `main` worker, which will still serve http requests, but at least TM won't be affecting other workers.

Long term, we'll look toward a strategy of having a dedicated process for background tasks to fully separate them from http requests.

without requiring an operator to scale out the server, as it would allow the server to handle
more http requests at once, making it less likely that a single bad request could delay the
event loop and impact subsequent requests.

The introduction of a clustering mode would allow spawning multiple Kibana processes ('workers')
from a single Kibana instance. (See [Alternatives](#8-alternatives) to learn more about the
difference between clustering and worker pools). You can think of these processes as individual
instances of the Kibana server which listen on the same port on the same machine, and serve
incoming traffic in a round-robin fashion.

Our intent is to eventually make clustering the default behavior in Kibana, taking advantage of
all available CPUs out of the box. However, this should still be an optional way to run Kibana
since users might have use cases for single-process instances (for example, users running Kibana
inside Docker containers might choose to rather use their container orchestration to run a
container per host CPU with a single Kibana process per container).

# 3. Architecture

> In 'classic' mode, the Kibana server is started in the main Node.js process.

Author: Main advantage I see would be memory consumption. Booting up a coordinator in 'classic' mode would potentially double the memory usage of the single worker cluster.

> In 'classic' mode, the Kibana server is started in the main Node.js process.

Reviewer 6: > Booting up a coordinator in 'classic' mode would potentially double the memory usage of the single worker cluster.

To attempt to mitigate this, we could determine whether to run the coordinator based on whether there are `> 1` configured workers, but this introduces more complexity:
1. Needs more complicated config validation: If there's only 1 `worker` configured, we'd ideally want to disallow configuring any `coordinator` options as they would be ignored.
2. If running a single worker (without a coordinator) that's configured in the yaml, we'd run into the same issue we discussed [in this thread](https://github.com/elastic/kibana/pull/94057#discussion_r645676628) with needing a way to read from the config before the process starts.

I'm also not sure it would make debugging any easier, as now we'd have implicitly different behaviors when there's 1 worker vs 2+.

Overall I think Josh's proposal here would be the much simpler implementation, but I agree that it may not be worth the memory tradeoff.

(which totally make sense, most of our requests response time is caused by awaiting ES response), 
the improvements of the clustering mode on performance under heavy load are far from negligible.

## 4.2 Testing against cloud

There is currently no easy way to test the performance improvements this could provide on Cloud, as we can't
deploy custom builds or branches on Cloud at the moment.

On Cloud, Kibana is running in a containerised environment using CPU CFS quota and CPU shares. 

If we want to investigate the potential perf improvement on Cloud further, our only option would be to setup a 
similar-ish environment locally (which wasn't done during the initial investigation).

# 5. Detailed design

## 5.1 Enabling clustering mode

Enabling clustering mode will be done using the `clustering.enabled` configuration property.

If clustering is enabled by default, then no configuration would be required by users, and
Kibana would automatically use all available cores. However, more detailed configuration
would be available for users with more advanced use cases:
```yaml
clustering:
  enabled: true               # enabled by default
  coordinator:

>     max_old_space_size: 1gb   # optional, allows to configure memory limit for coordinator only

Author: Sharing the same concerns here. Would requires to use an intermediary process to be able to read the config and start the coordinator with the correct var envs.

>     max_old_space_size: 1gb   # optional, allows to configure memory limit for coordinator only

Reviewer 6: > we have to set the NODE_OPTOONS environment variable before the process is started. Wouldn't we need some sort of bootstrapping process that reads config before starting the coordinator process?

Correct, I should have called this out specifically in the text. There are two caveats here:
- We'd need to read this config before we're fully bootstrapped
- We can't do config reloading on this item for the same reason; the process needs to be restarted for changes to take effect

The alternative is adding adding other environment variables to `node.options` that somehow map to configured workers, but when @kobelb and I discussed this, we liked the idea of keeping folks out of Node-specific configurations and centralizing things in the config as much as possible.

Open to other thoughts/ideas here though. Mostly I prefer this as a better DX, despite the fact that the implementation would require a little more work.

>     max_old_space_size: 1gb   # optional, allows to configure memory limit for coordinator only

Author: > We'd need to read this config before we're fully bootstrapped

More specifically, either reading it in shell before invoking `node` in our start script (not sure about the feasibility given our various distributions tbh), or by having the 'root' node process to bootstrap the coordinator process, instead of having it be the coordinator as done in current POC

>  Mostly I prefer this as a better DX, despite the fact that the implementation would require a little more work.

It would indeed be great to be able to configure that via the configuration file. Implementation may be complex enough to not justify the cost though.

>     max_old_space_size: 1gb   # optional, allows to configure memory limit for coordinator only

Author: > sort of like a "not ready" process

let's call it `preboot` process :trollface:

  workers:                    # when provided, total # of workers would be based on # of items in this list

>     - name: foo               # not technically necessary, but useful for logging purposes

Reviewer 3: Heh, I hadn't @pgayvallet's comment in support of this https://github.com/elastic/kibana/pull/94057#issuecomment-851444387

I'm not too opinionated either way, but I agree that we should be considering how to support node types on this same configuration.

As an aside, I wonder if `clustering` is going to make much sense to an end-user. Maybe we should use `node.workers` similar to Elasticsearch's configuration.

>     - name: foo               # not technically necessary, but useful for logging purposes

Reviewer 6: > As an aside, I wonder if clustering is going to make much sense to an end-user. Maybe we should use node.workers similar to Elasticsearch's configuration.

That's a good point - I like the idea of `node.workers` (or would we risk confusion with Node.js workers?)

Naming's the worst, would love to hear others' suggestions on this as well. I'm certainly open to the idea of another name.

> This might actually be nicer for debugging since I suspect we may expose the worker number in our API to plugins at some point.

I like your idea, I think it serves the same benefit of being future-proof as the proposed idea, but with the bonus of allowing for a more "basic" config

>     - name: foo               # not technically necessary, but useful for logging purposes

Reviewer 3: > or would we risk confusion with Node.js workers?

I don't think we should let the underlying technology that Kibana is built with shape how we name things for end-users too much. I imagine many sysadmins aren't also Node.js experts nor would they really need to care about how Kibana achieves multi-CPU utilization. I think `workers` is common vernacular and the consistency with Elasticsearch is valuable.

>     - name: foo               # not technically necessary, but useful for logging purposes

Reviewer 6: > I don't think we should let the underlying technology that Kibana is built with shape how we name things for end-users too much

Fair enough, that's also the same rationale for trying to allow configuring `max_old_space_size` via the yml and moving away from the `node.options` file, so it makes sense to me. And 100% agree the consistency with ES would be nice.

Hearing no other suggestions or objections, I'll go ahead and include this in the last batch of updates so we can move this into final review period.

>     - name: foo               # not technically necessary, but useful for logging purposes

Reviewer 6: Updated.

>     - name: foo               # not technically necessary, but useful for logging purposes

Reviewer 6: I renamed the core `clustering` service to the `node` service for consistency with the configs. 

The one thing I don't love with this rename is the `node.enabled` setting, which doesn't feel as intuitive (maybe `node.workers.enabled`?). @joshdover wdyt?

      max_old_space_size: 1gb # optional, allows to configure memory limits per-worker
    - name: bar
      max_old_space_size: 1gb
```

This per-worker design would give us the flexibility to eventually provide more fine-grained configuration,
like dedicated workers for http requests or background jobs.

## 5.2 Cross-worker communication

For some of our changes (such as the `/status` API, see below), we will need some kind of cross-worker
communication. This will need to pass through the coordinator, which will also serve as an 'event bus',
or IPC forwarder.

This IPC API will be exposed from the clustering service:

```ts
export interface ClusteringServiceSetup {
  // [...]
  broadcast: (type: string, payload?: ClusterMessagePayload, options?: BroadcastOptions) => void;
  addMessageHandler: (type: string, handler: MessageHandler) => MessageHandlerUnsubscribeFn;
}
```

To preserve isolation and to avoid creating an implicit cross-plugin API, handlers registered from a
given plugin will only be invoked for messages sent by the same plugin.

Notes:
-  To reduce clustered and non-clustered mode divergence, in non-clustered mode, these APIs would just be no-ops. 

>    It will avoid forcing (most) code to check which mode Kibana is running before calling them.

Reviewer 3: I'm assuming when `sendToSelf` is true that the message will still be sent, correct?

Reviewer 6: Ooh that's a good point, I agree in that case we'd probably still want to send the message.

> -  We could eventually use an Observable pattern instead of a handler pattern to subscribe to messages.

Reviewer 3: Still a little uneasy about not using Observables here. This would be one of the few (only?) Core event APIs that does not use Observables, however my pains with memory and performance overhead with Observables on the status API does make me a bit less cautious about this divergence here.

Reviewer 6: Yeah perf overhead was my main concern here. I guess the options are:
- We go with the existing proposal, evaluate how it's going, and decide whether to add observables as an enhancement later
- We drop the existing proposal and go with observables from the beginning, and see if we can track memory usage to determine whether this becomes a problem down the road

Thoughts? Anyone have strong feelings on this?

I think observables is more optimal from a design perspective, but the existing proposal feels less risky.

Author: See https://github.com/elastic/kibana/pull/94057#discussion_r620028092 for more details about the complexity of the observable based API

> I think observables is more optimal from a design perspective, but the existing proposal feels less risky.

My thoughts exactly. Would love to have an obs API, but implementation seems (way?) more complex to handle all the edge cases

Reviewer 3: Yep, I'm fine with this for now. We explore an Observable implementation later and evaluate the cost. Consistency in the Core APIs is really valuable IMO, but stability and performance takes precedent.

research would be required if this proved to be a necessity.

# 6. Technical impact

This section attempts to be an exhaustive inventory of the changes that would be required to support clustering mode.

## 6.1 Technical impact on Core

### 6.1.1 Handling multi-process logs

This is an example of log output in a 2 workers cluster, coming from the POC:

```
[2021-03-02T10:23:41.834+01:00][INFO ][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.840+01:00][INFO ][plugins-service] Plugin initialization disabled.
[2021-03-02T10:23:41.900+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
[2021-03-02T10:23:41.903+01:00][WARN ][savedobjects-service] Skipping Saved Object migrations on startup. Note: Individual documents will still be migrated when read or written.
```

The workers logs are interleaved, and, most importantly, there is no way to see which process each log entry is coming from. 
We will need to address that.

#### Options we considered:

1. Having a distinct logging configuration (with separate log files) for each worker

> 2. Centralizing log collection in the coordinator and writing all logs to a single file (or stdout)

Reviewer 6: Agreed, the idea is that this would be a subset of option 2. Basically the coordinator is still in charge of collecting/writing all logs, but will also update the default log pattern so worker IDs are written to the logs.

If the collector handles this, it doesn't really matter whether the worker is printing their own ID, or if the collector does it, as the collector already will know these IDs anyway.

#### Our recommended approach:

Overall we recommend keeping a single log file (option 2), and centralizing the logging system in the coordinator,
with each worker sending the coordinator log messages via IPC. While this is a more complex implementation in terms
of our logging system, it solves several problems:
- Preserves backwards compatibility.
- Avoids the issue of interleaved log messages that could occur with multiple processes writing to the same file or stdout.

> - Provides a solution for the rolling-file appender (see below), as the coordinator would handle rolling all log file

Reviewer 3: ```suggestion
- Provides a solution for the rolling-file appender (see below), as the coordinator would handle rolling all log files
```

> (small caveat being a slight loss in file size precision).

Reviewer 3: If only the coordinator is writing to disk, is this still true?

Reviewer 6: Oops, you're right. Got some threads crossed here, removed.

correct design, the main downside is that the logging implementation would be totally different in cluster and 
non cluster mode, and seems to be way more work that the other options.

#### Our recommended approach:
Even though it's more complex, we feel that centralizing the logging system in the coordinator is the right move here,
as it will also solve for how to enable the coordinator to log its own messages.

### 6.1.3 The status API

In clustering mode, the workers will all have an individual status. One could have a connectivity issue with ES 
while the other ones are green. Hitting the `/status` endpoint will reach a random (and different each time) worker, 
meaning that it would not be possible to know the status of the cluster as a whole.

We will need to add some centralized status state in the coordinator. Also, as the `/status` endpoint cannot be served 
from the coordinator, we will also need to have the workers retrieve the global status from the coordinator to serve 
the status endpoint.

We may also want to have the `/status` endpoint display each individual worker status in addition to the 

> global status, which may be breaking change in the `/status` API response format.

Reviewer 3: I think we can achieve both without breaking the API by having an overall status that shows the highest-severity status level for each component and the per-worker status as an extension of the existing response shape.

Reviewer 6: Ah, I missed updating this section, thanks

help validate this.

Note that if we did need to have per-worker UUIDs, this could be a breaking change, as the single `server.uuid`
configuration property would not be enough. If this change becomes necessary, one approach could be to have unique worker
IDs with `${serverUuid}-${workerId}`.

## 6.2 Technical impact on Plugins

### 6.2.1 What types of things could break?

#### Concurrent access to the same resources

Is there, for example, some part of the code that is accessing and writing files from the data folder (or anywhere else) 
and makes the assumption that it is the sole process actually writing to that file?

#### Using instanceUUID as a unique Kibana process identifier

Is there, for example, schedulers that are using the instanceUUID a single process id, in opposition to a single 
Kibana instance id? Are there situations where having the same instance UUID for all the workers is going to be a problem?

#### Things needing to run only once per Kibana instance

Is there any part of the code that needs to be executed only once in a multi-worker mode, such as initialization code, 
or starting schedulers? 

> An example would be Reporting's queueFactory pooling. As we want to only be running a single headless at a time per 

Reviewer 2: nit: 

```suggestion
An example would be Reporting's queueFactory polling. As we want to only be running a single headless at a time per 
```

Reviewer 9: Reporting uses Task Manager to queue reports and distribute their execution across multiple instances. TM task definitions have a property of [`maxConcurrency`](https://github.com/elastic/kibana/pull/90365), which we set to `1` for the `reporting:execute` task, to ensure only one job is executing in the instance at a time.

It's an implementation detail, but we will have to define what `maxConcurrency` tasks mean in clustering mode.

> Kibana instance, only one worker should have pooling enabled.

Reviewer 2: nit:
```suggestion
Kibana instance, only one worker should have polling enabled.
```

> 

Reviewer 9: > we want to only be running a single headless at a time per Kibana instance

If we didn't do anything, the number of Chromium processes running would always be <= to the number of cores. That seems OK to me.

### 6.2.2 Identified required changes

#### Reporting

We will probably want to restrict to a single headless per Kibana instance. For that, we will have to change the logic 
in [createQueueFactory](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/reporting/server/lib/create_queue.ts#L60-L64) 

> to only have the 'main' worker be pooling for reporting tasks.

Reviewer 2: nit:

```suggestion
to only have the 'main' worker be polling for reporting tasks.
```

- Tasks storing telemetry data
  
We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
tasks will be registered multiple times.

Note that sending the data multiple times doesn’t have any real consequences, apart from the additional number of ES requests, 
so this should be considered non-blocking and only an improvement.

- Event-based telemetry

Event-based telemetry may be affected as well. Both the existing one in the Security Solutions team and the general 
one that is in the works. More specifically, the size of the queues will be multiplied per worker, also growing in the
amount of network bandwidth used, and potentially affecting our customers.

We could address that by making sure that the queues are held only in the main worker.

#### Task Manager

Currently, task manager does "claims" for jobs to run based on the server uuid. We think this could still work with 
a multi-process setup - each task manager in the worker would be doing "claims" for the same server uuid, which
seems functionally the same as setting max_workers to `current max_workers * number of workers`.

> However, as a first step we can simply run Task Manager on the main worker. This doesn't completely solve potential

Reviewer 10: I would like to add an option to the table. The task manager doesn't need to be using the server uuid for claiming jobs, it just needs a unique identifier. If we can compose something like server_uuid + worker id, this would suffice as well because we can still trace tasks to servers while benefiting clustering if we wish to have each worker run task manager.

cc @gmmorris to keep me honest.

Reviewer 11: Yeah, 100% correct.
We use UUID to make it easier to map an owner ID on a  task back to a Kibana, but we can also generate a TM specific UUIS.

noisy neighbor problems as the main worker will still be receiving & serving http requests, however it will at least
ensure that other worker processes are free to serve http requests without risk of TM interference. Long term, we
could explore manually spawning a dedicated child process for background tasks that can be called from workers, and
thinking of a way for plugins to tell Core when they need to run things in the background.

It would be ideal if we could eventually solve this with our multi-process setup, however this needs more design work
and could necessitate an RFC in its own right. The key thing to take away here is that the work we are doing in this
RFC would not prevent us from exploring this path further in a subsequent phase. In fact, it could prove to be a
helpful first step in that direction.

#### Alerting

> Do we need the alerting task runner (`TaskRunner`, `TaskRunnerFactory`) to be running on a single worker?

Reviewer 10: Not that I can recall, it would inherit the same requirements as the task manager section.

Reviewer 10: I might have misread the question, is this question about having workers that only run alerting tasks? (nothing else)

Reviewer 9: @mikecote @gmmorris I think the question should be generalized wider than just alerting tasks. If a task is defined with a `maxConcurrency` option (which only Reporting at this moment is AFAIK), is the number of running tasks allowed still controlled at the Kibana instance level, or the worker level? 

I don't have an opinion either way. I think it will be OK for Reporting if the number of running tasks is limited by the number of cores.

Reviewer 11: Ah, this is a good point.
It's worker level, if I understand the terminology correctly.
Basically, if each core is running its own Task Manager, then there could easily be multiple reporting tasks running on the same machine (up to 1 per core).

## 6.3 Summary of breaking changes

### 6.3.1 status API

If we want to have the status API returns each individual worker's status, we will need to change the output of the 
status API in clustering mode. 

Note that the new format for /api/status is still behind a v8format flag, meaning that if we do these changes before 8.0, 
we won't be introducing any breaking change later.

### 6.3.2 stats API

> TODO

Reviewer 3: Seems similar to status endpoint in that we'll likely want to report per-worker stats. Which leads me to my next question: how should we be presenting this data in the Stack Monitoring UI? What changes do we need to coordinate with that team?

Reviewer 6: > What changes do we need to coordinate with that team?

This will depend on how we decide to move forward with `/stats` and the metrics service -- I just finished adding comments to that section (was the last one on my list).

Whether or not we end up introducing a breaking change, the metricbeat module will be affected as well.

is ES, maybe exposing a worker thread service and API from Core (task_manager would be a perfect example of potential consumer) 
would make more sense.

However, we believe the simplicity and broad acceptance of the `cluster` API in the Node community makes it the
better approach over `worker_threads`, and would prefer to only go down the road of a worker pool as a last resort.

Another alternative would be to provide tooling to ease the deployment of multi-instance Kibana setups, and only support
multi-instance mode moving forward.

# 9. Adoption strategy

Because the changes proposed in this RFC touch the lowest levels of Kibana's core, and therefore have potential to impact
large swaths of Kibana's codebase, we propse a multi-phase strategy:

## Phase 0
In the prepratory phase, we will evolve the existing POC to validate the finer details of this RFC, while also putting together
a more detailed testing strategy that can be used to benchmark our future work.

## Phase 1
To start implementation, we will make the required changes in Core, adding the `clustering.enabled` configuration property

> in development mode only. This way, we allow developers to test their features against clustering mode and to adapt their code

Reviewer 3: I suspect we'll need to allow this config in non-dev mode in order to run functional tests against this. We may even want a dedicated CI job on that runs master daily with the full test suite with clustering enabled.

We can throw a big warning log in there though so that it's super clear this shouldn't be used in production yet.

Reviewer 6: Yeah that makes sense, will update

ensure that other worker processes are free to serve http requests without risk of TM interference. Long term, we
could explore manually spawning a dedicated child process for background tasks that can be called from workers, and
thinking of a way for plugins to tell Core when they need to run things in the background.

It would be ideal if we could eventually solve this with our multi-process setup, however this needs more design work
and could necessitate an RFC in its own right. The key thing to take away here is that the work we are doing in this
RFC would not prevent us from exploring this path further in a subsequent phase. In fact, it could prove to be a
helpful first step in that direction.

#### Alerting

Currently haven't identified any Alerting-specific requirements that aren't already covered by the
Task Manager requirements.

## 6.3 Summary of breaking changes

### 6.3.1 instanceUUID

Depending on whether we are able to identify any cases where a shared uuid would be problematic, we may have to change
the server.uuid configuration property when clustering mode is enabled, which would be a breaking change.

> At this point, we have not identified any places where this is strictly necessary.

Reviewer 12: To make sure I understand it correctly: we don't need to introduce a breaking change. Do we?

Reviewer 6: > we don't need to introduce a breaking change

Correct. Alerting uses `server.uuid` currently, but we agreed we could introduce a new ID that combines `server.uuid` and `worker.id` for their use case.

Reviewer 6: I went ahead and removed this from the "breaking changes" section to prevent confusion.

### 6.3.2 `/stats` API & metrics service

The `process` memory usage reported here doesn't really make sense in a multi-process Kibana.
One option to consider is breaking the API by deprecating `process`, or turning it into a list of

> process stats reported from each worker. This is an open question still.

Reviewer 6: Yes, at a minimum we'd need to deprecate the old format. However, I would propose we make the breaking change in 8.0 because the aggregated data will be less precise and make less sense in a multi-process setup.

      "resident_set_size_bytes": 563625984
    },
    "pid": 52646,
    "event_loop_delay": 0.22967800498008728,
    "uptime_ms": 1706021.930404
  },
  // ...
}
```

As each request could be routed to a different worker, different results may come back each time.

This endpoint, registered from the `usage_collection` plugin, is getting these stats from Core's
`metrics` service (`getOpsMetrics$`), which is also used in the `monitoring` plugin for stats
collection.

While we could try to follow a plan similar to `/status` and have each worker report metrics to
the coordinator, there isn't a particularly sensible way to consolidate these. (e.g. Taking
the average of these values isn't really helpful in case there's a scenario where one process
is, for example, experiencing a spike in `event_loop_delay` while the others are operating
normally).

Ultimately we could extend the API to provide per-worker stats, but the question remains what we
should do with the existing `process` stats:
1. Deprecate them? (breaking change)
2. Accept a situation where they may be round-robined to different workers? (probably no)
3. Try to consolidate them somehow? (can't think of a good way to do this)

> 4. Always return stats for one process, e.g. main or coordinator? (doesn't give us the full picture)

Reviewer 6: @joshdover @pgayvallet This is the last open question I've run into, would appreciate your thoughts on it.

I don't love any of these options as I'd really like to avoid a breaking change if we can help it, but maybe you'll have some other ideas 🙂

Author: I think we want to do the same than we're planning to do for the `/status` endpoint / data, and allow to access per-worker process stats.

Now, for the current existing `OpsMetrics.process`, this is an excellent question, and I don't have any obvious answer. I do think none of the `2.`, `3.` or `4.` option really makes sense, but `1.` is introducing a breaking change, which doesn't play well at all with our goals on this matter, as we're very unlikely to finish the implementation before `8.0`. So, between `2.`, `3.` and `4.`, I would say that `4.` is (maybe?) the lesser evil

Reviewer 6: >  1. is introducing a breaking change, which doesn't play well at all with our goals on this matter, as we're very unlikely to finish the implementation before 8.0. So, between 2., 3. and 4., I would say that 4. is (maybe?) the lesser evil

These were my thoughts exactly, 4 feels like the least-bad option, but it's still not great.

> I think we want to do the same than we're planning to do for the /status endpoint / data, and allow to access per-worker process stats.

Yep, agreed on this too. I think we could solve for this in a variety of non-breaking ways. The ops metrics is the tricky part.

Reviewer 3: Whenever we make this feature GA, we'll going need to have Metricbeat updated to be able to pull data for multiple nodes against a single endpoint and/or have some way to display multiple processes in the Stack Monitoring UI. If there are any Metricbeat changes needed, I think it would make sense to prioritize getting those changes in early in the project so that very few users are flying blind or with only partial data during the eventual upgrade to clustering by default. If we were able to get these changes in for 7.last, then we'd be sure that this data wouldn't break in monitoring.

But we still should try to make sure that super old Metricbeat instances get something sorta sensible. I think aggregating the data for all processes (option 3) is probably going to create less headache than any other option. Even if it's not 100% accurate or sensible in this format.

Here's what I would propose as a way to aggregate each field:

https://github.com/elastic/kibana/blob/master/src/core/server/metrics/collectors/types.ts#L21-L42

```ts
export interface OpsProcessMetrics {
  /** process memory usage */
  memory: {
    /** heap memory usage */
    heap: {
      /** total heap available */
      total_in_bytes: number; // sum
      /** used heap */
      used_in_bytes: number; // sum
      /** v8 heap size limit */
      size_limit: number; // sum
    };
    /** node rss */
    resident_set_size_in_bytes: number; // sum
  };
  /** node event loop delay */
  event_loop_delay: number; // max?
  /** pid of the kibana process */
  pid: number; // coordinator process
  /** uptime of the kibana process */
  uptime_in_millis: number; // coordinator process
}
```

**heap.size_limit** is probably the one that makes the least sense because it would be possible for the graph to be showing it hasn't hit the limit, but a process may actually be very close to approaching the limit. There is some discussion about removing the heap limit from the graph though: https://github.com/elastic/kibana/issues/100408

Reviewer 6: @joshdover I think your proposal above makes sense:

- sum of available/used heap & node rss is straightforward
- `size_limit` sum could be confusing as you mention, but for the reasons explained in #100408, it's already creating confusion, so we aren't necessarily changing the status quo 🤷 
- `event_loop_delay` max makes sense to me. for debugging purposes we're mostly interested in that number if its high anyway. we could still discern from the logs which process has the high delay.
- `pid` and `uptime_in_millis` from the coordinator also make sense, especially with the initial implementation where we plan to kill all workers and coordinators if one of them dies. in the future if we respawn workers that die, this could be misleading, but hopefully by then we could have a solution in place on the Metricbeat side.

If we took this approach, the plan would then be:
- Use aggregated metrics in the initial implementation
- Add per-worker process metrics to the stats
  - Not sure exactly what this would look like -- ECS doesn't have a prescribed solution for this situation AFAICT -- but regardless we could do it in a non-breaking way
- Update Metricbeat module to use the new metrics format, and also Monitoring UI (hopefully by 7.last)

Long term we would still need to decide what to do about the existing `process`. Perhaps we could mark it as deprecated at least, even if we don't remove it in `8.0`. It wouldn't make sense to keep it around forever if the numbers could be misleading.

Reviewer 12: >  Perhaps we could mark it as deprecated at least, even if we don't remove it in 8.0. 

I find the option with aggregation to be a bit confusing for the end-users since they cannot grasp the state of the system at a single glance. 

My +1 to deprecate it, but keep it for some time. We can even consider outputting the process data in a different format for single-(`process: OpsProcessMetrics`) and multi-clustering modes (`process: OpsProcessMetrics[]`). Or use another end-point for `/stats` in multi-clustering mode.

Reviewer 6: I updated the language to clarify the plan: 

1. Deprecate `process` field & introduce aggregated metrics as an interim solution
2. Then later make the breaking changes, either removing `process` or supporting multiple formats as mentioned above

The good news is that we only really need step (1) to ship a first iteration of this, so we have some time to decide on the details around (2)
